app/gl_js.go

```diff
@@ -61,7 +63,21 @@ func (c *glContext) Lock() error {
 func (c *glContext) Unlock() {}
 
 func (c *glContext) Refresh() error {
-	return nil
+	if c.w.contextStatus == contextStatusOkay {
+		return nil
+	}
+	if c.w.contextStatus&contextStatusLost != 0 {
+		return errOutOfDate
+	}
+	if c.w.contextStatus&contextStatusDestroyed == 0 {
+		c.w.contextStatus |= contextStatusDestroyed
+		return gpu.ErrDeviceLost
+	}
+	if c.w.contextStatus&contextStatusRestored != 0 {
+		c.w.contextStatus = contextStatusOkay
+		return nil
+	}
+	return gpu.ErrDeviceLost
```

Comment:
That is chaotic, but the reason is: I decide to not change the `window.go` file, or add new errors, which would be exclusively for WASM.

--------

The `Refresh` is called when the context is already lost and a new GPU is created. Because of that, we have two issues:

1. It should not continue if the Status is Lost (that is unlike, since new frames are not generated).
2. It should destroy the current context (force Gio to recreate everything).
3. It should continue if the context is already destroyed (and then will be re-created).

Reply:
I don't see why you need contextStatusDestroyed. How about

```
	switch c.w.contextStatus {
	case contextStatusLost:
		return errOutOfDate
	case contextStatusRestored:
		c.w.contextStatus = contextStatusOkay
		return gpu.ErrDeviceLost
	default:
		return nil
	}
```

? If so, I believe you can demote context status from a bitfield to a state enum.